src/Util/Message.cc

```diff
@@ -0,0 +1,340 @@
+    moment_request.add_moments(CARTA::Moment::INTENSITY_WEIGHTED_DISPERSION_OF_THE_COORD);
+    moment_request.add_moments(CARTA::Moment::MEDIAN_OF_THE_SPECTRUM);
+    moment_request.add_moments(CARTA::Moment::MEDIAN_COORDINATE);
+    moment_request.add_moments(CARTA::Moment::STD_ABOUT_THE_MEAN_OF_THE_SPECTRUM);
+    moment_request.add_moments(CARTA::Moment::RMS_OF_THE_SPECTRUM);
+    moment_request.add_moments(CARTA::Moment::ABS_MEAN_DEVIATION_OF_THE_SPECTRUM);
+    moment_request.add_moments(CARTA::Moment::MAX_OF_THE_SPECTRUM);
+    moment_request.add_moments(CARTA::Moment::COORD_OF_THE_MAX_OF_THE_SPECTRUM);
+    moment_request.add_moments(CARTA::Moment::MIN_OF_THE_SPECTRUM);
+    moment_request.add_moments(CARTA::Moment::COORD_OF_THE_MIN_OF_THE_SPECTRUM);
+    return moment_request;
+}
+
+CARTA::EventType Message::EventType(std::vector<char>& message) {
+    carta::EventHeader head = *reinterpret_cast<const carta::EventHeader*>(message.data());
+    return static_cast<CARTA::EventType>(head.type);
+}
+
+void Message::FillHistogram(CARTA::Histogram* histogram, const carta::BasicStats<float>& stats, const carta::Histogram& hist) {
+    FillHistogram(histogram, hist.GetNbins(), hist.GetBinWidth(), hist.GetBinCenter(), hist.GetHistogramBins(), stats.mean, stats.stdDev);
+}
+
+void Message::FillHistogram(CARTA::Histogram* histogram, int num_bins, double bin_width, double first_bin_center,
+    const std::vector<int>& bins, double mean, double std_dev) {
+    if (histogram) {
+        histogram->set_num_bins(num_bins);
+        histogram->set_bin_width(bin_width);
+        histogram->set_first_bin_center(first_bin_center);
+        *histogram->mutable_bins() = {bins.begin(), bins.end()};
+        histogram->set_mean(mean);
+        histogram->set_std_dev(std_dev);
+    }
+}
+
+CARTA::SpectralProfileData Message::SpectralProfileData(int32_t file_id, int32_t region_id, int32_t stokes, float progress,
+    string& coordinate, vector<CARTA::StatsType>& required_stats, map<CARTA::StatsType, vector<double>>& spectral_data) {
+    CARTA::SpectralProfileData profile_message;
+    profile_message.set_file_id(file_id);
+    profile_message.set_region_id(region_id);
+    profile_message.set_stokes(stokes);
+    profile_message.set_progress(progress);
+
+    for (auto stats_type : required_stats) {
+        // one SpectralProfile per stats type
+        auto new_profile = profile_message.add_profiles();
+        new_profile->set_coordinate(coordinate);
+        new_profile->set_stats_type(stats_type);
+
+        if (spectral_data.find(stats_type) == spectral_data.end()) { // stat not provided
+            double nan_value = nan("");
+            new_profile->set_raw_values_fp64(&nan_value, sizeof(double));
+        } else {
+            new_profile->set_raw_values_fp64(spectral_data[stats_type].data(), spectral_data[stats_type].size() * sizeof(double));
+        }
+    }
+    return profile_message;
+}
+
+void Message::FillStatisticsValue(
```

Comment:
This should be called `FillStatisticsValues`, or just `FillStatistics`.

Reply:
My comment about the `FillHistogram` functions also applies to this. The `Fill...` functions should be outside the `Message` class (but still in the `Message` files).